[Merged by Bors] - feat(ring_theory/tensor_product): the base change of a linear map along an algebra

jcommelin opened the pull request:
---

- [x] depends on: #4769
- [x] depends on: #4771
- [x] depends on: #5506

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

jcommelin committed:
chore(linear_algebra/tensor_product): missing simp lemmas

jcommelin committed:
feat(linear_algebra/tensor_product): tensoring linear maps with modules

jcommelin committed:
feat(ring_theory/tensor_product): the base change of a linear map along an algebra

jcommelin committed:
generalize

rwbarton commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,234 @@
+variables [add_comm_monoid P] [semimodule R P]
+
+def assoc : ((M ⊗[A] N) ⊗[R] P) ≃ₗ[A] (M ⊗[A] (N ⊗[R] P)) :=
+_
```
I'm guessing this doesn't actually work?

kckennylau replied:
Done.

kckennylau committed:
revert direct limit content

kckennylau committed:
complete assoc

kckennylau committed:
Merge remote-tracking branch 'origin' into module-base-change

kckennylau committed:
fix errors

kckennylau committed:
lint

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,321 @@
+A linear equivalence constructing a linear map `M ⊗[R] N →[A] P` given a
+bilinear map `M →[A] N →[R] P` with the property that its composition with the
+canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is the given bilinear map `M →[A] N →[R] P`. -/
+def lift_equiv' : (M →ₗ[A] (N →ₗ[R] P)) ≃ₗ[A] ((M ⊗[R] N) →ₗ[A] P) :=
```
I think the other version is with a dot not an underscore?
```suggestion
/-- Heterobasic version of `tensor_product.lift.equiv`:

A linear equivalence constructing a linear map `M ⊗[R] N →[A] P` given a
bilinear map `M →[A] N →[R] P` with the property that its composition with the
canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is the given bilinear map `M →[A] N →[R] P`. -/
def lift.equiv' : (M →ₗ[A] (N →ₗ[R] P)) ≃ₗ[A] ((M ⊗[R] N) →ₗ[A] P) :=
```

kckennylau replied:
Done.

ChrisHughes24 commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,321 @@
+variables [semimodule R M] [semimodule R N]
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.restrict_scalars.semimodule R R M N =
```
Should this be in the tests folder?

jcommelin replied:
I would like to keep it here. It doesn't take up a lot of spacetime, and it is an important check that people reading/using this file might be worried about.

ocfnash replied:
Resolving as the instance and thus this example have now been removed.

ChrisHughes24 commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,321 @@
+  zero_smul := λ x, by simp only [rtensor_zero, alg_hom.map_zero, zero_apply] }
+
+/- Check that we didn't introduce a diamond. -/
+example : tensor_product.algebra_tensor_module R R M N =
```
Same comment here, should this be in the `tests` folder

ocfnash replied:
@jcommelin 's reply to an equivalent comment (now hidden) https://github.com/leanprover-community/mathlib/pull/4773#discussion_r526629293 was:
```
I would like to keep it here. It doesn't take up a lot of spacetime, and it is an important
check that people reading/using this file might be worried about.
```

ChrisHughes24 commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,321 @@
+variables (A)
+
+/-- `base_change A f` for `f : M →ₗ[R] N` is the `A`-linear map `A ⊗[R] M →ₗ[A] A ⊗[R] N`. -/
+def base_change (f : M →ₗ[R] N) : A ⊗[R] M →ₗ[A] A ⊗[R] N :=
```
Should this have type `(M →ₗ[R] N) →ₗ[R] (A ⊗[R] M →ₗ[A] A ⊗[R] N)`?

jcommelin replied:
Nope, because then you can no longer use dot.notation
But having such a bundled version is useful. So if it hasn't been added, then we should.

Vierkantor replied:
Could you add the linear map (as a new definition)?

kckennylau replied:
Done.

ChrisHughes24 reviewed (commented):
A few minor comments.

Vierkantor commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,321 @@
+open_locale tensor_product
+open tensor_product
+
+-- move this
```
Could you address this comment or is it no longer relevant?

kckennylau replied:
Done.

Vierkantor commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,321 @@
+@[simp] lemma lsmul_coe (a : A) : (lsmul R M a : M → M) = (•) a := rfl
+
+instance : is_scalar_tower R A A :=
+{ smul_assoc := λ r a b, by simp only [smul_eq_mul, smul_mul_assoc] }
```
This is [is_scalar_tower.right in ring_theory.algebra_tower](https://github.com/leanprover-community/mathlib/blob/master/src/ring_theory/algebra_tower.lean#L114), which should probably be moved higher in the import hierarchy if you need to redefine it here.

kckennylau replied:
Done.

Vierkantor commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,36 +1,321 @@
+  The code for
+  * `((A ⊗[R] B) ⊗[R] C) ≃ₐ[R] (A ⊗[R] (B ⊗[R] C))`
+  is written and compiles, but takes longer than the `-T100000` time limit,
+  so is currently commented out.
```
Is this `tensor_product.assoc` on line  241? It doesn't look commented out in the diff viewer.

kckennylau replied:
Done.

kckennylau committed:
per review

kckennylau committed:
Merge remote-tracking branch 'origin' into module-base-change

kckennylau committed:
line length

eric-wieser commented on src/algebra/algebra/tower.lean:
```diff
@@ -0,0 +1,281 @@
+by rw [algebra.algebra_map_eq_smul_one, smul_assoc, one_smul]
+
+variables {R A M}
+theorem smul_left_comm (r : R) (s : A) (x : M) : r • s • x = s • r • x :=
```
This should be changed to use `smul_comm_class`, right?

kckennylau replied:
Done.

eric-wieser commented on src/algebra/algebra/tower.lean:
```diff
@@ -0,0 +1,281 @@
+{ commutes' := λ _, (algebra_map_apply _ _ _ _).symm,
+  .. algebra_map S A }
+
+@[simp] lemma to_alg_hom_apply (y : S) : to_alg_hom R S A y = algebra_map S A y := rfl
```
I think this might be better as `to_alg_hom_coe`, equating  coe_fn's

kckennylau replied:
Done.

eric-wieser replied:
Although it might be worth keeping this as a non-simp lemma anyway

kckennylau replied:
I did.

kckennylau committed:
remove smul_left_comm

kckennylau committed:
coe_to_alg_hom

eric-wieser commented on src/algebra/algebra/tower.lean:
```diff
@@ -0,0 +1,282 @@
+    exact f.commutes (algebra_map R S r) },
+  .. (f : A →+* B) }
+
+@[simp] lemma restrict_base_apply (f : A →ₐ[S] B) (x : A) : restrict_base R f x = f x := rfl
```
This could probably be a coe lemma too

kckennylau replied:
Done.

kckennylau committed:
restrict_base

eric-wieser commented on src/algebra/algebra/tower.lean:
```diff
@@ -0,0 +1,286 @@
+
+variables {R S A}
+theorem of_algebra_map_eq [algebra R A]
+  (h : ∀ x, algebra_map R A x = algebra_map S A (algebra_map R S x)) :
```
This might make sense to express using `ring_hom.comp` rather than application, for similar reasons to the partially-applied ext lemmas library note

kckennylau replied:
Done in #5506.

eric-wieser commented on src/algebra/algebra/tower.lean:
```diff
@@ -0,0 +1,286 @@
+@[ext] lemma algebra.ext {S : Type u} {A : Type v} [comm_semiring S] [semiring A]
+  (h1 h2 : algebra S A) (h : ∀ {r : S} {x : A}, (by haveI := h1; exact r • x) = r • x) : h1 = h2 :=
+begin
+  unfreezingI { cases h1 with f1 g1 h11 h12, cases h2 with f2 g2 h21 h22,
```
Is there a shorter proof here from combining:
* `algebra.to_semimodule` is injective
* `semimodule.ext`, which was added recently

kckennylau replied:
Where is the lemma that `algebra.to_semimodule` is injective?

eric-wieser replied:
It probably doesn't exist, but I figured if we're adding a new lemma that might be a convenient one.

Or is this just moved code? It might make sense to split this PR in two, and do all the code moves separately.

eric-wieser commented on src/algebra/algebra/tower.lean:
```diff
@@ -0,0 +1,286 @@
+instance right : is_scalar_tower S A A :=
+⟨λ x y z, by rw [smul_eq_mul, smul_eq_mul, algebra.smul_mul_assoc]⟩
+
+instance nat : is_scalar_tower ℕ S A :=
```
Don't we already have this instance?

eric-wieser replied:
Oh, now I see these moved

kckennylau committed:
Merge remote-tracking branch 'origin' into module-base-change

eric-wieser commented on src/algebra/algebra/basic.lean:
```diff
@@ -1375,6 +1375,10 @@ instance is_scalar_tower_extend_scalars :
 
 variables {R S V W}
 
+lemma smul_apply' (c : S) (f : V →ₗ[R] W) (v : V) :
```
Can/should we change `smul_apply` to be this lemma, so that we don't need two?

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,68 +1,315 @@
+variables (R A M N)
+
+@[priority 500]
+instance restrict_scalars.semimodule : semimodule R (M ⊗[A] N) :=
```
Doesn't this already exist as `tensor_product.semimodule'`?

ocfnash replied:
What you say is almost true but a type synonym gets in the way. More precisely, the following fails with a complaint about not being able to find `[module R M]` and `[module R N]`:
```lean
def restrict_scalars.module : module R (M ⊗[A] N) :=
tensor_product.module'
```
but the following works:
```lean
def restrict_scalars.module : module R ((restrict_scalars R A M) ⊗[A] (restrict_scalars R A N)) :=
tensor_product.module'
```
I'm still trying to decide what the best setup is here.

(I realise this is out of date; these remarks are just to leave me with a clear conscience after I hit the resolve button.)

eric-wieser committed:
Merge branch 'master' into module-base-change

eric-wieser committed:
Merge branch 'master' into module-base-change

ocfnash committed:
Merge branch 'master' into module-base-change

ocfnash committed:
Replace `semimodule` --> `module`

ocfnash committed:
Easy fixes

ocfnash committed:
Temporarily comment out two defs that timeout

ocfnash committed:
Revert "Temporarily comment out two defs that timeout"

This reverts commit 4c99305bc41352f24c4612d7508d81d92e577daf.

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,307 @@
+@[priority 500]
+instance restrict_scalars.module : module R (M ⊗[A] N) :=
+show module R (restrict_scalars R A (M ⊗ N)), from
+restrict_scalars.module R A (M ⊗ N)
```
Reiterating [this hidden comment](https://github.com/leanprover-community/mathlib/pull/4773#discussion_r550436289): I think this instance already exists under `[is_scalar_tower R A M] [is_scalar_tower R A N]` assumptions, which probably means this instance is a bad idea.

ocfnash replied:
Noted, and this also looks suspicious to me. Looking at `linear_map.restrict_scalars`, I wonder do we need it?

ocfnash replied:
I have just added https://github.com/leanprover-community/mathlib/pull/4773/commits/89b22a27cddd615c21ef7fc228c1c5f55143e735 which removes this instance.

ocfnash committed:
Switch to `is_scalar_tower`-based approach for restriction of scalars

ocfnash committed:
Use type annotation instead of implicit variables for `curry'_inj`.

kbuzzard commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,292 @@
+variables [add_comm_monoid P] [module R P] [module A P] [is_scalar_tower R A P]
+
+variables {R A M N P}
+/-- Heterobasic version of `tensor_product.lift`:
```
I've never heard this word Heterobasic before. What does it mean?

ocfnash replied:
This is @jcommelin 's language. It was new to me too. I understand it to mean "mixed bases", using "base" in the sense of "base change" (i.e., two possibly different rings, in this affine case).

Vierkantor reviewed (approved):
Looks good to me, thanks!

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,292 @@
+  smul_add := λ a x y, by simp only [map_add],
+  smul_zero := λ a, by simp only [map_zero],
+  add_smul := λ a b x, by simp only [alg_hom.map_add, add_apply, rtensor_add],
+  zero_smul := λ x, by simp only [rtensor_zero, alg_hom.map_zero, zero_apply] }
```
If you remove the `nolint`ed `[distrib_mul_action R' N]` and `[module R'' N]` arguments from `tensor_product.has_scalar'`, then we already have this instance as `tensor_product.module'`:

https://github.com/leanprover-community/mathlib/blob/a741f646f0e3aac173f0ef6348f2cd1a2bf0a12e/src/linear_algebra/tensor_product.lean#L363-L366

The only reason why I added `@[nolint unused_arguments]` there in the first place was because I wasn't sure if this instance was a good idea as it's asymmetric, and thought I'd leave it until someone who knew more than me thought it was a good idea!

ocfnash replied:
Oh cool! Merging master to pick up the latest tweaks to `has_scalar'` and then I'll adjust accordingly.

ocfnash replied:
Now done.

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,292 @@
+
+lemma smul_def (a : A) (x : M ⊗[R] N) : a • x = (lsmul R M a).rtensor N x := rfl
+
+@[simp] lemma smul_tmul (c : A) (x : M) (y : N) : c • (x ⊗ₜ[R] y) = (c • x) ⊗ₜ[R] y := rfl
```
We already have this lemma as `tensor_product.smul_tmul'`, after you remove the `nolint` and clear up the lint errors:

https://github.com/leanprover-community/mathlib/blob/a741f646f0e3aac173f0ef6348f2cd1a2bf0a12e/src/linear_algebra/tensor_product.lean#L434-L438

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,292 @@
+@[ext] theorem ext {g h : (M ⊗[R] N) →ₗ[A] P}
+  (H : ∀ x y, g (x ⊗ₜ y) = h (x ⊗ₜ y)) : g = h :=
+linear_map.ext $ λ z, tensor_product.induction_on z (by simp_rw linear_map.map_zero) H $
+λ x y ihx ihy, by rw [g.map_add, h.map_add, ihx, ihy]
```
This would be a better ext lemma if you could state it as:
```
@[ext]
theorem curry'_inj {g h : M ⊗[R] N →ₗ[A] P}
  (H : curry' g = curry' h) : g = h :=
by {
  have := λ x, linear_map.congr_fun (linear_map.congr_fun H x),
  sorry -- your proof here
}
```

eric-wieser replied:
I think there's an easy proof of this using `linear_map.restrict_scalars_injective`; can you merge master so that that's available?

ocfnash replied:
Is this definitely a better ext lemma? I'll yield the point immediately if you still prefer it, as I'd love to get this merged.

Also, I just realised `curry'_inj` is already included in these changes: https://github.com/leanprover-community/mathlib/pull/4773/files#diff-40a1b7b0f1a7171e81d65a33e7c47bb428d521828f394e4bb5b4b9b70e82c657R142

In any case I've added https://github.com/leanprover-community/mathlib/pull/4773/commits/b97f3424a353241fa77147de84f808ec969f9170 with two lemmas I used in my proof:
```lean
begin
  suffices : curry (g.restrict_scalars R) = curry (h.restrict_scalars R),
  { exact linear_map.restrict_scalars_injective R (curry_inj this), },
  simp only [← restrict_scalars_curry', H],
end
```
before I realised `curry'_inj` already exists.

eric-wieser replied:
Yes, it's a better ext lemma because it handles a tower of towers correctly. For the same reason, `tensor_product.ext` is not actually marked `@[ext]`, and `mk_compr₂_inj` (aka curry_inj) is instead

eric-wieser replied:
The proof of `curry_injective` I was looking for is:
```lean
@[ext]
lemma curry'_injective : function.injective (curry' : (M ⊗ N →ₗ[A] P) → (M →ₗ[A] N →ₗ[R] P)) :=
λ x y h, linear_map.restrict_scalars_injective R $ curry_inj $
  (congr_arg (linear_map.restrict_scalars R) h : _)
```

eric-wieser replied:
Which gives you a shorter proof for `ext`:
```lean
theorem ext {g h : (M ⊗[R] N) →ₗ[A] P}
  (H : ∀ x y, g (x ⊗ₜ y) = h (x ⊗ₜ y)) : g = h :=
curry'_injective $ linear_map.ext₂ H
```

ocfnash replied:
Thanks, I'm actually working on this right now. Some fixes required after moving the `ext` but should be done shortly.

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,292 @@
+  begin
+    intros r a x,
+    simp only [algebra_tensor_module.smul_def, alg_hom.map_smul, smul_apply, rtensor_smul]
+  end }
```
Since the other comment got lost; #6741 adds a much more general version of this instance.

ocfnash replied:
I agree the #6741 instances dominate this one. I'm happy for that to be declared a dependent PR to merge before this, or to delegate the removal of this instance to a duty of that PR.

ocfnash replied:
After merging master (now containing #6741) I have removed this instance and one other.

ocfnash committed:
Merge branch 'master' into module-base-change

ocfnash committed:
Relax typeclass assumptions on `tensor_product.module'` and remove `algebra_tensor_module`

eric-wieser commented on src/linear_algebra/tensor_product.lean:
```diff
@@ -356,14 +356,23 @@ theorem smul.aux_of {R' : Type*} [has_scalar R' M] (r : R') (m : M) (n : N) :
+
+Note that in the special case that `R = R'`, since `R` is commutative, we just get the usual scalar
+action on a tensor product of two modules. This special case is important enough that, for
+perfomance reasons, we explicitly define it below, and reduce the priority of this instance. -/
```
```suggestion
performance reasons, we explicitly define it below, and reduce the priority of this instance. -/
```

eric-wieser commented on src/linear_algebra/tensor_product.lean:
```diff
@@ -356,14 +356,23 @@ theorem smul.aux_of {R' : Type*} [has_scalar R' M] (r : R') (m : M) (n : N) :
+Note that in the special case that `R = R'`, since `R` is commutative, we just get the usual scalar
+action on a tensor product of two modules. This special case is important enough that, for
+perfomance reasons, we explicitly define it below, and reduce the priority of this instance. -/
+@[priority 500]
```
It's not clear to me that there's any point reducing the priority here; we didn't need it before.

ocfnash replied:
Agreed. I actually just mindlessly copied this priority from the instance I deleted in the other file (`ring_theory/tensor_product.lean`) but when you put me on the spot I confess I cannot justify my claim. I'll remove these priorities.

eric-wieser commented on src/linear_algebra/tensor_product.lean:
```diff
@@ -356,14 +356,23 @@ theorem smul.aux_of {R' : Type*} [has_scalar R' M] (r : R') (m : M) (n : N) :
+(distributive) action of a second type of scalars `R'`, which commutes with the action of `R`, then
+the tensor product (over `R`) carries an action of `R'`.
+
+This instance defines this `R'` action in the case that it is the left module which has the `R'`
+action. Two natural ways in which this situation arises are:
```
I think it might be worth drawing attention to the fact that we have to choose between the left and right actions, and can't have both as this would cause typeclass diamonds.

ocfnash replied:
I think I'd rather keep silent on that point here as there is a common situation in which we have action on both factors, namely the tensor product of two representations, but then the action is different. E.g., for groups it's `g • (m ⊗ₜ n) = (g • m) ⊗ₜ (g • n)` so the story starts to get more complicated.

ocfnash replied:
Related remarks [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.234773.20base.20change/near/240929258)

ocfnash committed:
Remove dubious priorities.

ocfnash committed:
Fix long line

ocfnash committed:
Merge branch 'master' into module-base-change

ocfnash committed:
Bonus lemmas

eric-wieser commented on src/linear_algebra/tensor_product.lean:
```diff
@@ -632,6 +634,9 @@ def curry (f : M ⊗ N →ₗ P) : M →ₗ N →ₗ P := lcurry R M N P f
   curry f m n = f (m ⊗ₜ n) := rfl
 
+lemma curry_inj : function.injective (curry : (M ⊗[R] N →ₗ[R] P) → (M →ₗ[R] N →ₗ[R] P)) :=
+λ g h H, by { ext m n, simp only [linear_map.compr₂_apply, mk_apply, ← curry_apply, H], }
```
I think this statement is defeq to `mk_compr₂_inj`? So you can reuse the proof.

ocfnash replied:
Good call!

ocfnash committed:
Golf proof

ocfnash committed:
Appease linter

ocfnash committed:
Move ext attribute

ocfnash committed:
Move `curry`` to relax typeclass `[comm_semiring A]` to `[semiring A]`.

Without this `curry'_inj` actually applies less generally that `ext`.

eric-wieser commented on src/linear_algebra/tensor_product.lean:
```diff
@@ -632,6 +634,9 @@ def curry (f : M ⊗ N →ₗ P) : M →ₗ N →ₗ P := lcurry R M N P f
 @[simp] theorem curry_apply (f : M ⊗ N →ₗ[R] P) (m : M) (n : N) :
   curry f m n = f (m ⊗ₜ n) := rfl
 
+lemma curry_inj : function.injective (curry : (M ⊗[R] N →ₗ[R] P) → (M →ₗ[R] N →ₗ[R] P)) :=
```
The naming guide says this should be called `curry_injective`

ocfnash committed:
Eric's proof of `curry'_inj' ⛳️

ocfnash committed:
Move `ext` to get Eric's proof via `curry'_inj`

ocfnash committed:
Rename `curry_inj` and `curry'_inj` using suffix `_injective`

ocfnash committed:
Fix proofs broken by ext shuffling

ocfnash committed:
Simplify proof of `assoc`

Aside from now obeying the style guide, this is better because we
no longer have to invoke the underlying `ext` lemmas by name.

ocfnash committed:
Minor tidy ups

ocfnash committed:
Merge sections

ocfnash committed:
Fix long line

ocfnash committed:
Fix proof broken by new `ext` lemma

ocfnash committed:
Appease linter

ocfnash committed:
Merge branch 'master' into module-base-change

ocfnash committed:
Clean up after merging master

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,245 @@
+Constructing a linear map `M ⊗[R] N →[A] P` given a bilinear map `M →[A] N →[R] P` with the
+property that its composition with the canonical bilinear map `M →[A] N →[R] M ⊗[R] N` is
+the given bilinear map `M →[A] N →[R] P`. -/
+@[simps] def lift' (f : M →ₗ[A] (N →ₗ[R] P)) : (M ⊗[R] N) →ₗ[A] P :=
```
Given you're not in the `tensor_product` namespace here, should this just be called `lift` without the `'`?

ocfnash replied:
Agreed. I've renamed as follows the `algebra_tensor_module` namespace`:

- `lift'` → `lift`
- `lift.equiv'` → `lift.equiv`
- `mk'` → `mk`
- `curry'` → `curry`
- `lcurry'` → `lcurry`
- `uncurry'` → `uncurry`

ocfnash committed:
Remove trailing ' from definitions inside `algebra_tensor_module` namespace

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,247 @@
+
+lemma smul_eq_lsmul_rtensor (a : A) (x : M ⊗[R] N) : a • x = (lsmul R M a).rtensor N x := rfl
+
+/-- Heterobasic version of `tensor_product.curry`:
```
The statements below this one are true more generally for
```suggestion
end semiring

section semiring

variables [comm_semiring R] [semiring A] [has_scalar R A]
variables [add_comm_monoid M] [module R M] [module A M] [smul_comm_class R A M] [is_scalar_tower R A M]
variables [add_comm_monoid N] [module R N]
variables [add_comm_monoid P] [module R P] [module A P] [smul_comm_class R A P] [is_scalar_tower R A P]

/-- Heterobasic version of `tensor_product.curry`:
```

eric-wieser replied:
But maybe that's not interesting since it still requires `A` to be a semiring

ocfnash replied:
I agree this is probably the right statement. Indeed this could go in `linear_algebra/tensor_product.lean` and, I think, actually replace `tensor_product.curry` (via `smul_comm_class R R M` etc). The same goes for some other defs here.

I've shied away from aiming for greater generality here and elsewhere because I'm shooting to unblock extension of scalars and if I really step back and think about where we should eventually get, I see [quite a lot of work](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.234773.20base.20change/near/240929258).

What do you think? I'm certainly happy just to adjust this def as you suggest, see if I can thus replace `tensor_product.curry`, and leave it at that.

eric-wieser replied:
I think generalizing away `algebra` probably isn't all that useful here; the interesting generalizations I can think of require `distrib_mul_action_hom` instead of `linear_map`, and we don't have the theory of those developed (even though in all cases where both exist they're isomorphic).

> Indeed this could go in linear_algebra/tensor_product.lean and, I think, actually replace tensor_product.curry

This is quite tempting, but probably introduces some extra type arguments that would be annoying to pass explicitly. If we went down this route, I can imagine wanting `curry'` as your new def and `curry` as the less-complexly-typed version implemented with `curry`.

Perhaps the thing to do for now is add a comment somewhere in this file saying something along the lines of
```lean
/- As these heterobasic definitions are just more general versions of the definitions already in
`linear_algebra/tensor_product`, we could consider replacing the less general definitions with these ones.
If we do this, we probably should still implement the less general ones as abbreviations to the more general
ones with fewer type arguments. -/
````

ocfnash replied:
Agreed, and done.

ocfnash committed:
Fix after rename `curry'_apply` → `curry_apply`

eric-wieser commented on src/ring_theory/tensor_product.lean:
```diff
@@ -1,43 +1,247 @@
+open tensor_product
+
+/-!
+### The basechange of a linear map of `R`-modules to a linear map of `A`-modules
```
```suggestion
### The base-change of a linear map of `R`-modules to a linear map of `A`-modules
```

eric-wieser reviewed (approved):
bors d+

ocfnash committed:
Suggestions from CR
